proposals/0532-type-variable-scoping.rst

```diff
@@ -0,0 +1,161 @@
+But warnings must always be phrased in terms "that thing you did you might not have wanted to".
+That means making the student aware of the thing after all --- "that thing you did" is something the student was never taught and therefore should never be goaded by a warning into learning about after all.
+
+Concretely, for educational purposes we want to get rid of implicit bindings, and get rid of the *concept* of implicit binding.
+We want the student to be *unable* to write them, and we want the compiler to *not* tell them with other configuration options feature exists, the same way a Haskell 98 users should not be told about "type families" or weather.
+
+Single namespace syntactic uniformity
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+`#270`_ has a very nice story about making single name-space code not fork-like by accepting fewer programs.
+``-Werror=pattern-signature-binds`` either breaks the "non fork-like" condition, or breaks the "single namespace" condition.
+
+For example, suppose a single namespace user wrote::
+
+  import Foo (B) -- from `type B = Int`
+
+  a = Int
+
+  f (x :: a) = ....
+  f (x :: B) = ....
+
+``x :: B`` definitely doesn't bind ``B``, but what about ``x :: a``?
+
+#. If ``a`` is a use, then if this code is copy pasted to another file without ``-Wpuns -Wpun-bindings`` it will *change meaning**::
+
+     f (x :: Int) = ... -- no puns, after inlining
+     f = let type a = _ in \(x :: a) -> ... -- yes puns
+
+   This makes the pun change fork-like, and disqualifies it from being a warning.
+   That is exactly the sort of outcome many people are worried about "no pun idem" code leading to, and which the authors of `#270`_ have worked strenuously to avoid.
+
+#. If ``a`` is a bind, the meaning with and without puns will be the same::
+
+     f = let type a = _ in \(x :: a) -> ... -- no puns
+     f = let type a = _ in \(x :: a) -> ... -- yes puns
+
+   but the "illusion" of a single namespace is shattered, because ``a`` and ``B`` are *not* treated the same way.
+
+   Furthermore, as a practical matter, how *should* ``a`` be used here?
+   ``-Werror=pattern-signature-binds`` will catch and error on our pattern signature bind, but it won't allow us to refer to ``a``.
+   Not being able to use without rebinding it leaves an awkward tripping point that must be worked around with more variable indirection.
+
+With ``-XNoImplicitBinds`` instead, we go with option 2 to ``-Wpuns`` and ``-Wpun-binding`` stay genuine warnings that do not change behavior.
+``-XImplicitBinds`` however *can* change behavior, and it results in the same behavior (unification variable) regardless of ``-Wpuns`` and ``-Wpun-binding``.
+Documentation can make clear that, yes, the punning warnings are insufficient on their own to create the single-namespace "illusions", but that are intended to work with ``-XNoImplicitBinds`` in which case they do succeed in their intended purpose.
```

Comment:
Hm, good point that `-Wpuns` and `-Wpun-bindings` are insufficient to maintain the illusion of a single namespace. I have not thought about this. But saying that `-XNoImplicitBinds` is required would be a step too far. I believe we just need a third warning: `-Wterm-variable-capture`. Good news is that it is already part of an accepted proposal (namely [#281](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0281-visible-forall.rst)) and there's an implementation in [!9420](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/9420).


Your example

```
a = Int
f (x :: a) = ...
```

would trigger `-Wterm-variable-capture`, whereas something like

```
f (x :: b) = ...   -- no term-level ‘b’ in scope
```

would continue to work under `-Wterm-variable-capture` (but not under `-XNoImplicitBinds`).

Reply:
I think `-Wpun-bindings` would imply `-Wterm-variable-capture`? But sure `-XNoImplicitBinds` doesn't need to be formally required.

Basically the "issue" is that before there was term variable namespace lookup in types, `-XImplicitBinds` was purely monotonic in allowing more programs, but now it is not monotonic in that some valid programs will have their meaning changed. That need not be a show-stopper however, after all, extensions like `-XScopedTypeVariables` also change the meaning of valid programs and are thus not monotonic.

Reply:
Yes, I agree with your characterization of the issue around `-XImplicitBinds`, but you need not bring `-Wpuns` and `-Wpun-bindings` into this. Perhaps say that `-XImplicitBinds` may break programs that assume a single namespace, but `-Wterm-variable-capture` is available to detect such programs.

Reply:
https://github.com/Ericson2314/ghc-proposals/blob/type-variables/proposals/0448-type-variable-scoping.rst#unified-namespacing-and-extension-monotonicity has this!